contracts/src/Teleporter/upgrades/TeleporterUpgradeable.sol

```diff
@@ -56,16 +44,95 @@ abstract contract TeleporterUpgradeable {
 
         teleporterRegistry = TeleporterRegistry(teleporterRegistryAddress);
-        minTeleporterVersion = teleporterRegistry.latestVersion();
+        _minTeleporterVersion = teleporterRegistry.latestVersion();
+    }
+
+    /**
+     * @dev See {ITeleporterReceiver-receiveTeleporterMessage}
+     * Requirements:
+     *
+     * - `msg.sender` must be a Teleporter version greater than or equal to `minTeleporterVersion`.
+     */
+    function receiveTeleporterMessage(
+        bytes32 originBlockchainID,
+        address originSenderAddress,
+        bytes calldata message
+    ) external {
+        // Checks that `msg.sender` matches a Teleporter version greater than or equal to `minTeleporterVersion`.
+        require(
+            teleporterRegistry.getVersionFromAddress(msg.sender) >=
+                _minTeleporterVersion,
+            "TeleporterUpgradeable: invalid Teleporter sender"
+        );
+
+        _receiveTeleporterMessage(
+            originBlockchainID,
+            originSenderAddress,
+            message
+        );
     }
 
     /**
-     * @dev This is a virtual function that should be overridden to update the `minTeleporterVersion`
-     * allowed for modifier `onlyAllowedTeleporter`, and emit {MinTeleporterVersionUpdated} event after.
+     * @dev Updates the minimum Teleporter version allowed for delivering Teleporer messages
+     * to this contract.
      *
-     * Note: To prevent anyone from being able to call this function, which would disallow messages
+     * To prevent anyone from being able to call this function, which would disallow messages
      * from old Teleporter versions from being received, this function should be safeguarded with access
-     * controls. For example, if the derived contract has an owner/admin, only they can call this function.
+     * controls. This is done by overriding the implementation of {_checkTeleporterUpgradeAccess}.
+     */
+    function updateMinTeleporterVersion(uint256 version) public virtual {
+        _checkTeleporterUpgradeAccess();
+        _setMinTeleporterVersion(version);
+    }
+
+    /**
+     * @dev Public getter for `_minTeleporterVersion`.
+     */
+    function getMinTeleporterVersion() public view returns (uint256) {
+        return _minTeleporterVersion;
+    }
+
+    /**
+     * @dev Sets the minimum Teleporter version allowed for delivering Teleporter messages.
+     * Emits a {MinTeleporterVersionUpdated} event if the minimum Teleporter version was updated.
+     * Requirements:
+     *
+     * - `version` must be less than or equal to the latest Teleporter version.
+     * - `version` must be greater than the current minimum Teleporter version.
+     *
+     */
+    function _setMinTeleporterVersion(uint256 version) internal virtual {
+        uint256 latestTeleporterVersion = teleporterRegistry.latestVersion();
+        uint256 oldMinTeleporterVersion = _minTeleporterVersion;
+
+        require(
+            version <= latestTeleporterVersion,
```

Comment:
Instead of checking that `version <= latestTeleporterVersion`, should we consider explicitly requiring that `getAddressFromVersion(version) != address(0)`? I believe this is a more restrictive check.

Reply:
I think it's a design decision but I think I lean slightly towards the current implementation here.

Imagine we have a min version of 2, latest version of 4, but skipped version 3 entirely. The current implementation lets us set the min version to 3, which might be preferable in the case where someone wants to disallow version 2, but is fine with version 3 if it is later added to the registry.

This might cause bugs elsewhere here though, if we assume that the min version should exist. We should document this very clearly.

Reply:
I am also leaning towards allowing the dapp to update to a version that is not yet deployed, but within `minTeleporterVersion` and `latestVersion`.